### ISSUE TITLE
resolvectl can report dnssec state better

### ISSUE THREAD

**Issue by pemensik** [systemd/systemd#17226]
**Is your feature request related to a problem? Please describe.**
> [RFC 4035](https://tools.ietf.org/html/rfc4035) describes required changes for DNSSEC compatibility. It was made to keep compatibility with previous implementations, which never seen any DNSSEC records. Because systemd-resolved is more recent implementation, it should never need to behave like older implementations.

**Describe the solution you'd like**
Current naming of 'DNSSEC supported' is quite confusing. Frankly, there is no need to disable "support" for DNSSEC. Current software should just know it can be expected and how to handle it.

Instead, it should be possible to toggle only validation on and off. Even better, it should use negative trust anchors to disable only local domains overrides, where validation is turned off only for local domains, where DNSSEC fails for administration/policy reasons (router., internal. and similar domains). It should include more details how to add [trust anchor](https://www.freedesktop.org/software/systemd/man/dnssec-trust-anchors.d.html#), the best would be link from DNSSEC=.

- 'DNSSEC setting:' should become 'DNSSEC validation:'
- 'DNSSEC supported: yes' should be always builtin, no need to write it. Make sure it passes recommendation of [DNS flag day 2019](https://dnsflagday.net/2019/#experts), or [RFC 6891](https://tools.ietf.org/html/rfc6891#section-7). It should work with both DNSSEC aware and unaware servers. Allow non-default workarounds for common cases if required, but have to configure them. Instead, it should indicate whether servers on each link support dnssec or not, for example by -t NS . query to them first.
- DNSSEC= config option should be renamed to Validation=, reusing old value for backward compatibility.
- dnssec-trust-anchors.d man page should contain at least one concrete negative trust anchor example, for example 'test' domain.
- /usr/lib/dnssec-trust-anchors.d/ should be shipped by the package and contain built-in trust anchors in file(s). Should give users an example to copy & modify from.

**Describe alternatives you've considered**
[configuration manual](https://www.freedesktop.org/software/systemd/man/resolved.conf.html#DNSSEC=) is not very clear how it works and why would I need DNSSEC=no.

**Comment by pemensik** [comment]
Similar proposal is also offered in [RFC 8027, section 5.1](https://tools.ietf.org/html/rfc8027#section-5.1)

**Comment by pemensik** [comment]
Support for manually enabled outgoing queries should not be necessary. Just process correct detection of EDNS0 and DNSSEC support in 'upstream' resolvers. It it replies without EDNS or DO bit, mark it on the internal server record and don't send DO queries to it and disable local validation, if not better server available. If it replies FORMERR or any other negative answer, try the same. If passed this time, mark the internal server record again. Retry after some (configurable at best) amount of hours. Such algorithms work quite well almost every implementation. Few broken should be reported. Only timeout handling should support also quite broken forwarders without any proper response on DNSSEC query.

Global disabling of outgoing EDNS queries should be quite rare and the most of recent DNS servers does not support it for that reason. Only support for proper negotiation is required.

**Comment by keszybz** [comment]
> **Describe the solution you'd like**
> Current naming of 'DNSSEC supported' is quite confusing. Frankly, there is no need to disable "support" for DNSSEC. Current software should just know it can be expected and how to handle it.

The support for DNSSEC is an optional feature. Just like many other features, it can be enabled or disabled at compilation time, similarly to resolved itself. In particular systemd is often compiled for embedded systems where people carefully select only the features needed in specialized environments. 
 
>     * 'DNSSEC setting:' should become 'DNSSEC validation:'

I think we should rework this whole output. It's become unwieldy with the growing number of options. Something like +DNSSEC +LLMNR -mDNS would be much more compact without being any less readable.
 
>     * 'DNSSEC supported: yes' should be always builtin, no need to write it.

No, as mentioned above.

>     * DNSSEC= config option should be renamed to Validation=, reusing old value for backward compatibility.

Nah, DNSSEC is super clear at first sight, while "validation" not so much.

>     * dnssec-trust-anchors.d man page should contain at least one concrete negative trust anchor example, for example 'test' domain.

Sure, will add.

>     * /usr/lib/dnssec-trust-anchors.d/ should be shipped by the package and contain built-in trust anchors in file(s). Should give users an example to copy & modify from.

Nah, not really. NTA is not something that most users will need to tweak. Having the NTAs directly in the code make things faster and simpler to package. No need to make things more complicated.

> **Describe alternatives you've considered**
> [configuration manual](https://www.freedesktop.org/software/systemd/man/resolved.conf.html#DNSSEC=) is not very clear how it works and why would I need DNSSEC=no.

See above.

**Comment by pemensik** [comment]
> 
> The support for DNSSEC is an optional feature. Just like many other features, it can be enabled or disabled at compilation time, similarly to resolved itself. In particular systemd is often compiled for embedded systems where people carefully select only the features needed in specialized environments.

There are different different support levels for DNSSEC. Very basic DNSSEC support has no reason to be optional feature. DNSSEC support means ability to receive and forward RRSIG records, EDNS0 extension and NSEC3 records. Just receive them and cache. No crypto library is needed for that. I see no reason to omit it in even on embedded system.

Then DNSSEC validation requires understanding of DNSSEC keys, RRSIG format and crypto support. I very much agree *VALIDATION* is optional. That was the reason for my request.
> 
> > ```
> > * 'DNSSEC setting:' should become 'DNSSEC validation:'
> > ```
> 
> I think we should rework this whole output. It's become unwieldy with the growing number of options. Something like +DNSSEC +LLMNR -mDNS would be much more compact without being any less readable.

It should be able to output, which forwarder supports EDNS0 and which supports DNSSEC. Then it should indicate whether validation is done for those replies.
> 
> > ```
> > * DNSSEC= config option should be renamed to Validation=, reusing old value for backward compatibility.
> > ```
> 
> Nah, DNSSEC is super clear at first sight, while "validation" not so much.

No, it is not. You proved it by confusing support for DNSSEC records with ability for validating they are unmodified and signature is valid. Those are different things. I think the difference is not understood by systemd-resolved team, so it cannot be uderstood by their users.
> 
> > ```
> > * dnssec-trust-anchors.d man page should contain at least one concrete negative trust anchor example, for example 'test' domain.
> > ```
> 
> Sure, will add.
> 
> > ```
> > * /usr/lib/dnssec-trust-anchors.d/ should be shipped by the package and contain built-in trust anchors in file(s). Should give users an example to copy & modify from.
> > ```
> 
> Nah, not really. NTA is not something that most users will need to tweak. Having the NTAs directly in the code make things faster and simpler to package. No need to make things more complicated.

No. Because you tend to give advice to your users to turn off whole DNSSEC support. They would usually need just working negotiation inside systemd (take advantage it contains state) and negative trust anchors. It is simple to waive it but hard to make it working properly. If you struggle with packaging one file and directory, I can help with that.

**Comment by keszybz** [comment]
> There are different different support levels for DNSSEC. Very basic DNSSEC support has no reason to be optional feature. DNSSEC support means ability to receive and forward RRSIG records, EDNS0 extension and NSEC3 records. 

That makes sense. But if we implement the missing support for forwarding DNSSEC-related records, and this support will be "free" in the sense of only requiring a bit of code and it will not be disable it at compile time, there will be no reason to report this. So the current switch which reports if dnssec *validation* is compiled in and enabled is appropriate.

> Because you tend to give advice to your users to turn off whole DNSSEC support.

We don't. We allow it to be disabled. There's really nothing special about this. If you look at the configure output during build, it has >100 top-level features that can be enabled and disabled at will.

> It should be able to output, which forwarder supports EDNS0 and which supports DNSSEC. Then it should indicate whether validation is done for those replies.

Well, pretty much everyone should support EDNS0 now. And DNSSEC validation is reported alrady, as explained above.

**Comment by pemensik** [comment]
> That makes sense. But if we implement the missing support for forwarding DNSSEC-related records, and this support will be "free" in the sense of only requiring a bit of code and it will not be disable it at compile time, there will be no reason to report this. So the current switch which reports if dnssec _validation_ is compiled in and enabled is appropriate.

I agree. But that is a reason, why I requested renaming it from DNSSEC support to DNSSEC validation. It should be clear, that only validation is toggled. Not support compiled-in, whatever its state is.

> We don't. We allow it to be disabled. There's really nothing special about this. If you look at the configure output during build, it has >100 top-level features that can be enabled and disabled at will.

I can point to a  [message](https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/message/P63RI3VBQ7NGL3AKMTR7PCVHVSCPYCLF/), where instead of suggesting NTA on *fritz.box*, turning off DNSSEC "solved it". It were not you, but there is issue for it here on the project. You maybe didn't, but someone did.

> 
> > It should be able to output, which forwarder supports EDNS0 and which supports DNSSEC. Then it should indicate whether validation is done for those replies.
> 
> Well, pretty much everyone should support EDNS0 now. And DNSSEC validation is reported alrady, as explained above.

Unfortunately not yet. Some old Mikrotik routers still do not support even EDNS. DNSSEC "support" is reported. But that does not show whether a forwarder can reply on DO=1 query with DO=1 response. Which some home routers still don't support, making *validation* impossible, even when turned on. It would be nice to see if DNSSEC is possible on current network, even when not enabled (yet).

**Comment by keszybz** [comment]
> I agree. But that is a reason, why I requested renaming it from DNSSEC support to DNSSEC validation.

fe37e5a5d1 changed the output to not say "support". And the documentation talks about "DNSSEC validation", so I think we're OK now.

> I can point to a message, where instead of suggesting NTA on fritz.box, turning off DNSSEC "solved it".

This message was discussing issues with endpoint routers.

> Unfortunately not yet. Some old Mikrotik routers still do not support even EDNS. 

Right, but how is that relevant to resolvectl output about systemd-resolved capabilities? resolvectl is only about resolved, it cannot communicate with some arbitrary router. Or maybe you mean that we should report for each listed server what capabilities is has? I guess that could be useful in some circumstances, but I think the logs are a better place for this... I don't think we want to build a dbus api to expose such itsy bitsy pieces of information.

**Comment by pemensik** [comment]
> [fe37e5a](https://github.com/systemd/systemd/commit/fe37e5a5d192ec55f87cd57893688a865b7f72d2) changed the output to not say "support". And the documentation talks about "DNSSEC validation", so I think we're OK now.

Almost. DNSSEC client support is still missing, but better. Thank you.
 
> 
> > Unfortunately not yet. Some old Mikrotik routers still do not support even EDNS.
> 
> Right, but how is that relevant to resolvectl output about systemd-resolved capabilities? resolvectl is only about resolved, it cannot communicate with some arbitrary router. Or maybe you mean that we should report for each listed server what capabilities is has? I guess that could be useful in some circumstances, but I think the logs are a better place for this... I don't think we want to build a dbus api to expose such itsy bitsy pieces of information.

Oh, it is obtained by D-bus, it makes sense to not fetch support flags for each resolver. It seems DNSSEC supported: shows features of a router. systemd-resolved purpose is to communicate with configured DNS resolvers. I am not proposing scanning that router. Just note flags in request and in reply match. Logging detected features might be useful, I think logs are enough. I think it is already done.

Is there help or description, what values of DNSSEC supported mean? I think *yes* and *no* are very clear. Could it just report supported or unsupported with allow-downgrade setting? I know something about DNSSEC, but still I am unsure what it means. Could it be called "autodetect" to be less confusing? Would be *downgraded* better word than unsupported, if allow-downgrade is used?

**Comment by pemensik** [comment]
Also I think DNSOverTLS is also not pure boolean, but has oppportunistic parameter. It should show actual state, which might be dynamic. I think it should be similar to DNSSEC state, which has similar allow-downgrade option. It could even use the same code, if *autodetect* were called state between *yes* and *no*.

**Comment by poettering** [comment]
I think there's some conceptual misunderstanding here. You appear to consider systemd-resolved to be a DNS server primarily. It isn't really. It's a client, with primary APIs being glibc NSS (via nss-resolve), D-Bus and Varlink. 

Now, there are some local clients that do not support either of the three, and insist to speak raw DNS. For those cases we do support the local stub, but that's a minimum support thing, with a clear lockdown for local clients. It's thus still not a real DNS server, but just a client that has a compat ifaces that does the minimum stuff that local clients that insist on speaking only DNS can use.

Now, if we have accepted now that resolved is a *client* to DNS primarily, and not so much a *server*, I think the "DNSSEC=" setting it has is pretty clear too: it's about whether it consumes the DNSSEC records from its upstream servers (i.e. validates with them) or not. It is not about being a DNSSEC server to its own clients, simply because the relevant getaddrinfo()/NSS APIs do not expose anything like that. Yes, the DNS stub can expose it (and with #17535 we finally do), but philosophically that still doesn#t make us a real DNS server, it makes us a client with compat support.

Anyway, I am not convinced we should rename anything from DNSSEC to DNSSEC validaiton. It's clear enough.

**Comment by poettering** [comment]
EDNS0 support is not as established as you imply, for example we still have #8608 (which ubuntu patches in downstream, and I am inclined to merge upstream too). 

Given that resolved is primarily a *client* I think it's essential for us to focus on what the accepted feature set of DNS servers actual clients talk to is. And that means the DNS servers cheap wifi routers expose, and not the authoritative DNS servers that server some domains on the internet. The latter might be fine with EDNS0 always, the former certainly still aren't. And the "edns0 flagday" stuff doesn't really matter for them either. You won't find regular people buying a router just because some dns dude decided that something is a flag day. We can't just break DNS for them, we *have* to do our best to handle that.

**Comment by poettering** [comment]
Anyway, is there anything left open here that you wanted to see fixed, @keszybz? Otherwise I think we can close this?

**Comment by pemensik** [comment]
> I think there's some conceptual misunderstanding here. You appear to consider systemd-resolved to be a DNS server primarily. It isn't really. It's a client, with primary APIs being glibc NSS (via nss-resolve), D-Bus and Varlink.
It receives queries from other subsystems. It might not be classic authoritative or recursive server, but it clearly is NOT a client. If you would open /etc/resolv.conf, the most useful word there is *nameserver*. It states it clearly, considers it server. Clients consider it as well. If it should not act like server, just leave that file to appropriate configuration.
> 
> Now, there are some local clients that do not support either of the three, and insist to speak raw DNS.
They speak DNS, because they send queries to DNS server. For whatever reason they do, either act as server or direct them to a real server.

> For those cases we do support the local stub, but that's a minimum support thing, with a clear lockdown for local clients. It's thus still not a real DNS server, but just a client that has a compat ifaces that does the minimum stuff that local clients that insist on speaking only DNS can use.
> 
> Now, if we have accepted now that resolved is a _client_ to DNS primarily, and not so much a _server_, I think the "DNSSEC=" setting it has is pretty clear too: it's about whether it consumes the DNSSEC records from its upstream servers (i.e. validates with them) or not. It is not about being a DNSSEC server to its own clients, simply because the relevant getaddrinfo()/NSS APIs do not expose anything like that.

The discussion was about domain port 53 handled by systemd-resolved, right? So no, you are repeating it is the client and it still is not. I am not pushing to expose DNSSEC via getaddrinfo(). If you want to implement nscd replacement and provide cache just for getaddrinfo, be my guest. If all DNS queries are redirected to domain port, it has to handle correct DNS protocol with up-to-date extensions and pass them forward.

It is not pretty clear. You can consume DNSSEC records but not validate them, if upstream server provides them. Or you can do validation on them, ensuring cache is protected against cache poisoning. Because it did not provide DNSSEC records to its own clients, the only state available were in resolvectl status. So it could mean multiple things without clear description, what exactly it means.

> Yes, the DNS stub can expose it (and with #17535 we finally do), but philosophically that still doesn#t make us a real DNS server, it makes us a client with compat support.
Oh thank you! It make resolved server with DNSSEC support. Quite similar to *dnsmasq* (check ``dnf info dnsmasq``), systemd-resolved is a cache server, which is not full recursive server, but close to it.
> 
> Anyway, I am not convinced we should rename anything from DNSSEC to DNSSEC validaiton. It's clear enough.
It could mean:
1. Upstream (link) server(s) can provide DNSSEC
2. Validation is done on received DNSSEC records
3. DNSSEC is offered to local clients

I don't think it makes it clear, which case it is. I think (1) can now be reported by just a query to 127.0.0.53 with DNSSEC, but still might be reported. Only (2) should be configured, because only this might be disabled. (3) should be handled by negotiation well.

**Comment by pemensik** [comment]
> EDNS0 support is not as established as you imply, for example we still have #8608 (which ubuntu patches in downstream, and I am inclined to merge upstream too).

Oh don't! This fix is clearly wrong and should be applied only after manual configuration or to known names list. It MUST not be aplied to all hosts. It would duplicate number of queries to all nonexistent names, just because some broken implementation in a bus or train. This MUST not be used for any NXDOMAIN. NXDOMAIN is clear signal no such name exists. IT should be the same under any combination of flags in request. It should be reported to network owner instead. It might make sense to have it configured per-connection. But it should make it clear to user, this is wrong and should be reported (at least to the log)

> Given that resolved is primarily a _client_ I think it's essential for us to focus on what the accepted feature set of DNS servers actual clients talk to is. And that means the DNS servers cheap wifi routers expose, and not the authoritative DNS servers that server some domains on the internet. The latter might be fine with EDNS0 always, the former certainly still aren't. And the "edns0 flagday" stuff doesn't really matter for them either. You won't find regular people buying a router just because some dns dude decided that something is a flag day. We can't just break DNS for them, we _have_ to do our best to handle that. No workaround from dns-violations should be applied without user intervention. It is named violation for no reason.

I admit cheap wifi box support for any advanced DNS queries if often terrible. But most of them can work just fine, often breaking DNSSEC incoming from client. In contrast to that, recursive resolvers, to which are queries from clients sent, usually handle any extensions according to specs.

You can't break them? I have Mikrotik router at home. And it is broken under systemd-resolved, because it provides name ``router.`` with its dynamic address. Such address does not resolve to anything, because it does not support LLMNR. And because you refuse to send single label to configured DNS server. DNS flag day was about both recursive and authoritative servers. Unfortunately, automated probes can check authoritative servers and they did. They were not able to test cheap boxes at home, because no route there were found. Many cheap boxes ignores extensions, but that is okay.

Don't break it for them (thanks again for DNSSEC support!), but provide only manual fixes in situations with quite broken implementation like #8608. Provide feedback to user, so he don't have to buy broken router again.

**Comment by keszybz** [comment]
I think we implemented everything that was contained in the original list, or decided not to implement it. Let's close this.
